client/src/compute/rest/index.ts

```diff
@@ -0,0 +1,135 @@
+import { clearTokens, getAccessToken } from "./auth";
+import { ComputeState, getApiConfig } from "../common";
+
+export interface Config {
+  endpoint: string;
+  clientId?: string;
+  clientSecret?: string;
+  context?: string;
+  serverId?: string;
+}
+
+import { ComputeServer } from "../server";
+import { ComputeSession } from "../session";
+import { ComputeJob } from "../job";
+
+let config: Config;
+let computeSession: ComputeSession | undefined;
+
+async function setup() {
+  const apiConfig = getApiConfig();
+  if (!config.serverId) {
+    apiConfig.accessToken = await getAccessToken(config);
+  }
+
+  if (computeSession && computeSession.sessionId) {
+    const state = await computeSession
+      .getState()
+      .catch(() => (computeSession = undefined));
+    if (state === ComputeState.Error) {
+      await computeSession.cancel();
+    }
+  }
+
+  if (computeSession) return;
+
+  //Set the local in the base options so it appears on all api calls
+  const locale = JSON.parse(process.env.VSCODE_NLS_CONFIG ?? "{}").locale;
+  apiConfig.baseOptions["headers"] = { "Accept-Language": locale };
+
+  if (config.serverId) {
+    const server1 = new ComputeServer(config.serverId);
+    computeSession = await server1.getSession();
+
+    //Maybe wait for session to be initialized?
+    return;
+  } else {
+    //Create session from context
+    const contextsApi = ContextsApi(apiConfig);
+    const contextName = config.context || "SAS Job Execution compute context";
+    const context = (
+      await contextsApi.getContexts({
+        filter: `eq(name,'${contextName}')`,
+      })
+    ).data.items[0];
+    if (!context?.id)
+      throw new Error("Compute Context not found: " + contextName);
+
+    const sess = (
+      await contextsApi.createSession(
+        { contextId: context.id },
+        { headers: { "accept-language": locale } }
+      )
+    ).data;
+    computeSession = ComputeSession.fromInterface(sess);
+  }
+}
+
+/*
+Prints the job log in an async manner.
+*/
+async function printLog(job: ComputeJob, onLog?: (logs: LogLine[]) => void) {
+  const log = await job.getLogStream();
+  for await (const line of log) {
+    onLog([line]);
+  }
+}
+
+async function run(code: string, onLog?: (logs: LogLine[]) => void) {
+  if (!computeSession?.sessionId) throw new Error();
+
+  //Get the job
+  const job = await computeSession.execute({ code: [code] });
+  let state = await job.getState();
+
+  const retLog = printLog(job, onLog);
+
+  //Wait until the job is complete
+  do {
+    state = await job.getState({ onChange: true, wait: 2 });
+  } while (await job.isDone(state));
+
+  //Clear out the logs
+  await retLog;
+
+  //Now get the results
+  const results = await job.results();
+
+  const res: RunResult = {
+    html5: "",
+    title: "",
+  };
+  //We can return more than just html, but for right now we are only returning
+  //the first HTML file that we get
+  for (const result of results) {
```

Comment:
In current main branch, it will get the "last" rather than "first" result. For example user closed the default ods and started a new one.

Reply:
I figured as much. I was working on a change where ALL the ODS results were returned and multiple WebView's were opened, but I thought that change was too much for this PR so I left it out. I will change it back to using the last one.